mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala

```diff
@@ -442,11 +488,17 @@ object RFormulaModel extends MLReadable[RFormulaModel] {
       val metadata = DefaultParamsReader.loadMetadata(path, sc, className)
 
       val dataPath = new Path(path, "data").toString
-      val data = sparkSession.read.parquet(dataPath).select("label", "terms", "hasIntercept").head()
+      val df = sparkSession.read.parquet(dataPath)
+      val columns = Seq("label", "terms", "hasIntercept", "evalExprs")
+      val data = df.select(columns.intersect(df.columns).map(df(_)): _*)
+        .head()
       val label = data.getString(0)
       val terms = data.getAs[Seq[Seq[String]]](1)
       val hasIntercept = data.getBoolean(2)
-      val resolvedRFormula = ResolvedRFormula(label, terms, hasIntercept)
+      // Check for old saved models (spark version < 3), they do not have evalExprs.
+      val (major, minor) = VersionUtils.majorMinorVersion(metadata.sparkVersion)
+      val evalExprs = if (major.toInt < 3) Seq[String]() else data.getAs[Seq[String]](3)
+      val resolvedRFormula = ResolvedRFormula(label, terms, hasIntercept, evalExprs)
```

Comment:
Looks correct. We better have a test for this. We can load a model from 2.4 in a test case. There are few examples in ml.

Reply:
Thanks, added a testcase for loading a model from 2.4.4.